server/src/main/java/com/epam/aidial/core/server/controller/ControllerSelector.java

```diff
@@ -339,57 +271,76 @@ private static Controller selectPost(Proxy proxy, ProxyContext context, String p
+        });
+        delete(PATTERN_RESOURCE, (proxy, context, pathMatcher) -> {
             ResourceController controller = new ResourceController(proxy, context, false);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
-
-        match = match(INVITATION, path, context);
-        if (match != null) {
-            String invitationId = UrlUtil.decodePath(match.group(1));
+        });
+        delete(INVITATION, (proxy, context, pathMatcher) -> {
+            String invitationId = UrlUtil.decodePath(pathMatcher.group(1));
             InvitationController controller = new InvitationController(proxy, context);
             return () -> controller.deleteInvitation(invitationId);
-        }
-
-        return null;
-    }
-
-    private static Controller selectPut(Proxy proxy, ProxyContext context, String path) {
-        Matcher match = match(PATTERN_FILES, path, context);
-        if (match != null) {
+        });
+        // PUT routes
+        put(PATTERN_FILES, (proxy, context, pathMatcher) -> {
             UploadFileController controller = new UploadFileController(proxy, context);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
-
-        match = match(PATTERN_RESOURCE, path, context);
-        if (match != null) {
+        });
+        put(PATTERN_RESOURCE, (proxy, context, pathMatcher) -> {
             ResourceController controller = new ResourceController(proxy, context, false);
+            String path = context.getRequest().path();
             return () -> controller.handle(resourcePath(path));
-        }
+        });
+    }
 
-        return null;
+    public ControllerSelection select(HttpServerRequest request) {
+        HttpMethod method = request.method();
+        String path = request.path();
+        if (request instanceof HttpServerRequestInternal req) {
+            ControllerSelection selection = req.context().getLocal(CTX_KEY);
+            if (selection == null) {
+                selection = select(method, path);
+                req.context().putLocal(CTX_KEY, selection);
+            }
+            return selection;
+        }
+        return select(method, path);
     }
 
-    private Matcher match(Pattern pattern, String path, ProxyContext context) {
-        Matcher matcher = pattern.matcher(path);
-        if (matcher.find()) {
-            SpanUtil.updateName(pattern, path, context.getRequest().method().name());
-            return matcher;
-        }
-        return null;
+    public ControllerSelection select(String path) {
+        return select(null, path);
+    }
+
+    private ControllerSelection select(HttpMethod method, String path) {
+        return ROUTES.stream()
+                .map(r -> r.select(method, path))
+                .filter(Objects::nonNull)
+                .max(Comparator.comparing(ControllerSelection::pathTemplate))
```

Comment:
Why do we need to call max? Perhaps just findAny()?

Reply:
fixed